### PR TITLE
Adds support for encoded filters for API use

### LINKED ISSUE #555: add `filter` QSP to API requests that accepts a base64 encoded filter payload
It won't always be possible to send a JSON payload with `POST` to a filter endpoint. For example, when we want a browser to handle a result or a response natively that is generated from a link or a form, only form multipart data (for POSTs) or query string parameters (for GET/POST) are supported.

If you're in this situation can send a filter body, as per the above specification, encoded in the `filter` query string parameter.

- query string parameter name: `filter`
- format: base64 encoded JSON payload of a standard filter object
- precedence: the payload from the `filter` parameter has the lowest priority. Parameters from other query string parameters or the body of a filter request will override the keys from the `filter` payload.

### PR DESCRIPTION
**Pull request by atruskie**
# Adds support for encoded filters for API use.

The API will now accept a filter encoded in base64url from either the query string or from body parameters.

This allows a simpler invocation of a filter endpoint via a GET request.

Fixes #555.

## Changes

Adds support for decoding filters from the request parameter `filter_encoded`. A decoded filter is deep-merged into the hash of other request parameters.

Wiki also updated with new documentation: https://github.com/QutEcoacoustics/baw-server/wiki/API:-Filtering#advanced-filtering-with-a-query-string-parameter

## Problems

Rails 7 and Ruby 3.1 upgrades bundled in. You can ignore those changes in your review (just review the last commit).

## Visual Changes

N/A

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Remove/Reduce warnings from edited files
- [x] Unit tests have been added for changes
- [ ] Ensure CI build is passing

### CODE REVIEW

File: spec/lib/modules/filter/query_filter_encoded_spec.rb [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+          }).query_full
+        }.to raise_error(
+          CustomErrors::FilterArgumentError,
+          /filter_encoded was not a valid JSON payload: unexpected token at '.*'/
+        )
+      end
+    end
+
+    context 'when decoding' do
+      it 'works as intended' do
```
Comment (Allcharles) [inline spec/lib/modules/filter/query_filter_encoded_spec.rb:79]: Would like to see non latin1 characters used in the test text as well (ie. russian language or symbols)

File: spec/requests/cms_spec.rb [hunk 2]
```diff
@@ -1,13 +1,11 @@
...
 
-
-
 describe 'CMS' do
   create_standard_cms_pages
   prepare_users
 
   # assets should always be available
-  context 'allows users with no credentials to fetch assets' do
+  context 'when users have no credntials, the can fetch assets' do
```
Comment (Allcharles) [inline spec/requests/cms_spec.rb:8]:
```suggestion
context 'when users have no credentials, they can fetch assets' do
```

**Review by Allcharles — APPROVED** [review]
Looks good, I much enjoy the smaller PRs. Makes it easier to track whats being done. Although I would have preferred this split into two separate PRs, however i suspect they were a package deal by the time you realised

File: app/modules/filter/query.rb [hunk 3]
```diff
@@ -217,6 +219,35 @@ def has_filter_params?
...
+      rescue StandardError
+        raise CustomErrors::FilterArgumentError, 'filter_encoded was not a valid RFC 4648 base64url string'
+      end
+
+      hash = begin
+        JSON.parse(json)
+      rescue StandardError => e
+        message = e.message.gsub(/\d+: /, '')
+        raise CustomErrors::FilterArgumentError, "filter_encoded was not a valid JSON payload: #{message}"
+      end
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/filter/query.rb:240]: Yeah, there was no fixed limit for length support. It varies with every piece of software between browser and Rails.

Instead, the best I could is just check the JSON payload is valid.

[the remark this answers is not in the document]
Comment (Allcharles) [reply app/modules/filter/query.rb:240]: Does the error message include the payload with some regex sanitation?

[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/filter/query.rb:240]: Doesn't include payload, does include JSON parser error, 

Regex sanitization is to remove a useless integer error code.

**Pull request merged**